samples/snippets/src/main/java/com/example/bigquerystorage/WriteToDefaultStream.java

```diff
@@ -191,7 +197,37 @@ public void onFailure(Throwable throwable) {
             return;
           } catch (Exception e) {
             // Fall through to return error.
-            System.out.format("Failed to retry append: %s\n", e);
+            System.out.format("Failed to retry append: %s%n", e);
+          }
+        }
+
+        if (throwable instanceof AppendSerializtionError) {
+          AppendSerializtionError ase = (AppendSerializtionError) throwable;
+          Map<Integer, String> rowIndexToErrorMessage = ase.getRowIndexToErrorMessage();
+          if (rowIndexToErrorMessage.size() > 0) {
+            // Omit the faulty rows
+            JSONArray dataNew = new JSONArray();
+            for (int i = 0; i < appendContext.data.length(); i++) {
+              if (!rowIndexToErrorMessage.containsKey(i)) {
+                dataNew.put(appendContext.data.get(i));
+              } else {
+                // process faulty rows by placing them on a dead-letter-queue, for instance
+              }
+            }
+
+            // Mark the existing attempt as done since we got a response for it
+            done();
+
+            // Retry the remaining valid rows.
+            if (dataNew.length() > 0) {
+              try {
+                this.parent.append(new AppendContext(dataNew, 0));
```

Comment:
Is there a way to start another thread and do this append? Currently this append can block waiting for inflight limit...

Reply:
Done. Both retries and row-errors are now handled by calls to append() made from a separate thread.